### PR TITLE
New interface with XX handshake

### PR DESCRIPTION
**Pull request by morrigan**
Interface according to: https://github.com/libp2p/js-interfaces/tree/master/src/crypto

and integration of handshake using XX

### CODE REVIEW

File: src/handshake.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { bytes, bytes32 } from "./types/basic";
+import { NoiseSession, XXHandshake } from "./xx";
+import { KeyPair } from "./types/libp2p";
+
+export class Handshake {
+  static async runXX(
```
Comment (mpetrunic) [inline src/handshake.ts]: I would convert this to constructor and have methods like init/exchange/finish like in secio

Comment (mpetrunic) [reply src/handshake.ts]: later on it would allow us to combine multiple handshakes

File: src/index.ts [hunk 2]
```diff
@@ -1,8 +1,5 @@
-import { encrypt } from './encrypt';
-
-const tag = '/noise';
+import { Noise } from './noise';
 
 export {
-  tag,
-  encrypt,
+  Noise,
```
Comment (mpetrunic) [inline src/index.ts]: export * from "./noise"

File: src/noise.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        privateKey: staticNoiseKey,
+        publicKey,
+      }
+    } else {
+      // todo: generate new static key
+    }
+  }
+
+  public protocol() {
+    return '/noise';
```
Comment (mpetrunic) [inline src/noise.ts]: I think this should be `public protocol = "/noise";`

File: src/types/libp2p.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+
+// Also seen as Pair
+export type Stream = {
+  sink(source: Iterable<any>),
+  source: Object,
+}
+
+export type Duplex = [Stream, Stream];
```
Comment (mpetrunic) [inline src/types/libp2p.ts]: this belongs in it-peer or something so you have to do 

```javascript
declare module "it-peer" {
    export type Duplex = [Stream, Stream];
}
```

File: src/types/libp2p.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  stats: ConnectionStats,
+
+  streams(): Duplex,
+  addStream(muxedStream: any) : Stream,
+}
+
+export interface NoiseConnection {
+  remoteEarlyData?(): bytes,
+  secureOutbound(insecure: InsecureConnection, remotePeer: PeerId): Promise<SecureConnection>,
+  secureInbound(insecure: InsecureConnection): Promise<SecureConnection>,
```
Comment (mpetrunic) [inline src/types/libp2p.ts]: I don't think that's correct types.
Secio returns just {conn, remotePeer}

Comment (morrigan) [reply src/types/libp2p.ts]: Mhm, will update according to https://github.com/libp2p/js-interfaces/tree/master/src/crypto#ap

**Comment by morrigan** [comment]
@mpetrunic I think we can merge so it doesn't get long. Still missing encryption/decryption after handshake

File: src/xx.ts [hunk 6]
```diff
@@ -272,6 +277,7 @@ export class XXHandshake {
   }
 
   private async readMessageA(hs: HandshakeState, message: MessageBuffer) : Promise<bytes> {
+    console.log("publci key: ", message.ne)
```
Comment (mpetrunic) [inline src/xx.ts]: remove console. You should probably add npm package `debug` and log stuff (separate PR probably)

**Pull request merged**